pkg/services/ngalert/schedule/schedule.go

```diff
@@ -653,7 +637,43 @@ func (r *alertRuleRegistry) keyMap() map[models.AlertRuleKey]struct{} {
 type alertRuleInfo struct {
 	evalCh  chan *evalContext
 	stopCh  chan struct{}
-	version int64
+	mtx     sync.Mutex
+	stopped *atomic.Bool // indicates that the stopCh was closed and therefore the rule evaluation routine should exit
+}
+
+func newAlertRuleInfo() *alertRuleInfo {
+	return &alertRuleInfo{evalCh: make(chan *evalContext), stopCh: make(chan struct{}), stopped: atomic.NewBool(false)}
+}
+
+// Stop signals the rule evaluation routine to stop via stopCh and sets flag stopped to true. Does nothing If the loop is stopped
+func (a *alertRuleInfo) Stop() bool {
```

Comment:
In the past I've just done something like:

```suggestion
func (a *alertRuleInfo) Stop() bool {
     a.mtx.Lock()
     defer a.mtx.Unlock()
     if !a.stopped {
          close(a.stopCh)
          a.stopped = true
     }
     return a.stopped
}
```